controllers/operands/wasp.go

```diff
@@ -0,0 +1,225 @@
+package operands
+
+import (
+	"errors"
+	"os"
+	"reflect"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hcov1beta1 "github.com/kubevirt/hyperconverged-cluster-operator/api/v1beta1"
+	"github.com/kubevirt/hyperconverged-cluster-operator/controllers/common"
+	hcoutil "github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
+)
+
+const (
+	pullURL = "quay.io/fdeutsch/wasp-operator-prototype:latest"
+	// For simplicity matters, re-use the virt-handler SA. That will spare
+	// the HCO controller from deploying custom privileged SA with SCC.
+	waspServiceAccount = "kubevirt-handler"
+)
+
+func newWaspHandler(Client client.Client, Scheme *runtime.Scheme) Operand {
+	return &conditionalHandler{
+		operand: &genericOperand{
+			Client: Client,
+			Scheme: Scheme,
+			crType: "DaemonSet",
+			hooks:  &waspHooks{},
+		},
+		getCRWithName: func(hc *hcov1beta1.HyperConverged) client.Object {
+			return NewWaspWithNameOnly(hc)
+		},
+		shouldDeploy: func(hc *hcov1beta1.HyperConverged) bool {
+			return hc.Spec.FeatureGates.EnableHigherDensityWithSwap != nil &&
+				*hc.Spec.FeatureGates.EnableHigherDensityWithSwap
+		},
+	}
+}
+
+type waspHooks struct {
+	cache *appsv1.DaemonSet
+}
+
+func (h *waspHooks) getFullCr(hc *hcov1beta1.HyperConverged) (client.Object, error) {
+	if h.cache == nil {
+		h.cache = NewWasp(hc)
+	}
+
+	return h.cache, nil
+}
+
+func (*waspHooks) getEmptyCr() client.Object { return &appsv1.DaemonSet{} }
+
+func (*waspHooks) updateCr(
+	req *common.HcoRequest,
+	Client client.Client,
+	exists runtime.Object,
+	required runtime.Object) (bool, bool, error) {
+	wasp, ok1 := required.(*appsv1.DaemonSet)
+	found, ok2 := exists.(*appsv1.DaemonSet)
+
+	if !ok1 || !ok2 {
+		return false, false, errors.New("can't convert to DaemonSet")
+	}
+
+	if !hasCorrectDaemonSetFields(found, wasp) {
+		overwritten := false
+		if req.HCOTriggered {
+			req.Logger.Info("Updating existing wasp DaemonSet Spec to new opinionated values")
+		} else {
+			req.Logger.Info("Reconciling an externally updated wasp DaemonSet Spec to its opinionated values")
+			overwritten = true
+		}
+		hcoutil.MergeLabels(&wasp.ObjectMeta, &found.ObjectMeta)
+		wasp.Spec.DeepCopyInto(&found.Spec)
+		err := Client.Update(req.Ctx, found)
+		if err != nil {
+			return false, false, err
+		}
+		return true, overwritten, nil
+	}
+
+	return false, false, nil
+}
+
+func (*waspHooks) justBeforeComplete(_ *common.HcoRequest) { /* no implementation */ }
+
+func (*waspHooks) getConditions(cr runtime.Object) []metav1.Condition {
+	return dsConditionsToK8s(cr.(*appsv1.DaemonSet).Status.Conditions)
+}
+
+func NewWasp(hc *hcov1beta1.HyperConverged) *appsv1.DaemonSet {
+	waspImage, _ := os.LookupEnv(hcoutil.WaspImageEnvV)
```

Comment:
please validate this env var actually exists and has a value before even creating the handler.

Reply:
Good point. Can be done when *creating* the handler, into a new field in the handler, then you could use this new field without checking the env var each time.

Reply:
@orenc1 @nunnatsa I have added:
```
	if ci.IsOpenshift() && ci.IsWaspImageProvided() {
		operands = append(operands, newWaspHandler(client, scheme))
	}
```
and implementation:
```
	waspImageValue, waspImageExists := os.LookupEnv(WaspImageEnvV)
	c.waspImageProvided = waspImageExists && len(waspImageValue) > 0
...
...
...
func (c *ClusterInfoImp) IsWaspImageProvided() bool {
	return c.waspImageProvided
}

```

Reply:
@nunnatsa @tiraboschi Hi. To panic makes sense to me as well, but I also tried to be consistent with the way the console plugin is being deployed. If we agree on panic I can defer it to another PR so we won't drift with the current one which is big enough already.